toolkit/tools/internal/packagerepo/repocloner/rpmrepocloner/rpmrepocloner.go

```diff
@@ -69,325 +92,266 @@ const (
 
-		bindFsType = ""
-		bindData   = ""
+	timestamp.StartEvent("cloning packages", nil)
+	defer timestamp.StopEvent(nil)
 
-		chrootLocalRpmsDir      = "/localrpms"
-		chrootLocalToolchainDir = "/toolchainrpms"
+	if len(packagesToClone) == 0 {
+		logger.Log.Debug("No packages to clone.")
+		return
+	}
 
-		overlaySource                  = "overlay"
-		overlayUpperDirectoryRpms      = "/overlaywork/upper_rpms"
-		overlayWorkDirectoryRpms       = "/overlaywork/workdir_rpms"
-		overlayUpperDirectoryToolchain = "/overlaywork/upper_toolchain"
-		overlayWorkDirectoryToolchain  = "/overlaywork/workdir_toolchain"
-	)
+	packageNames := make([]string, len(packagesToClone))
+	index := 0
+	for _, pkg := range packagesToClone {
+		packageNames[index] = convertPackageVersionToTdnfArg(pkg)
+		index++
+	}
 
-	r.usePreviewRepo = usePreviewRepo
-	if usePreviewRepo {
-		logger.Log.Info("Enabling preview repo")
+	depsSwitch := "--nodeps"
+	if cloneDeps {
+		depsSwitch = "--alldeps"
 	}
 
-	if disableDefaultRepos {
-		logger.Log.Info("Disabling default upstream PMC repositories")
+	constantArgs := []string{
+		"install",
+		"-y",
+		depsSwitch,
+		"--downloadonly",
+		"--downloaddir",
+		r.chrootCloneDir,
 	}
 
-	// Ensure that if initialization fails, the chroot is closed
-	defer func() {
-		if err != nil {
-			logger.Log.Warnf("Failed to initialize cloner. Error: %s", err)
-			if r.chroot != nil {
-				closeErr := r.chroot.Close(leaveChrootFilesOnDisk)
-				if closeErr != nil {
-					logger.Log.Panicf("Unable to close chroot on failed initialization. Error: %s", closeErr)
-				}
-			}
+	// Setting install root to a temp directory to prevent TDNF
+	// from skipping over packages already installed in the chroot.
+	// The 'reinstall' command doesn't support the '--nodeps' flag.
+
+	packagesCount := len(packageNames)
+	logger.Log.Debugf("Will clone in total %d packages.", packagesCount)
+
+	for i := 0; i < packagesCount; i += maxBatchSize {
+		batchEndIndex := i + maxBatchSize
+		if batchEndIndex > packagesCount {
+			batchEndIndex = packagesCount
 		}
-	}()
 
-	// Create the directory to download into
-	err = os.MkdirAll(destinationDir, os.ModePerm)
-	if err != nil {
-		logger.Log.Warnf("Could not create download directory (%s)", destinationDir)
-		return
-	}
+		currentPackageNames := packageNames[i:batchEndIndex]
 
-	// Setup the chroot
-	logger.Log.Infof("Creating cloning environment to populate (%s)", destinationDir)
-	r.chroot = safechroot.NewChroot(tmpDir, isExistingDir)
+		finalArgs := make([]string, len(constantArgs)+len(currentPackageNames))
+		finalArgs = append(constantArgs, currentPackageNames...)
 
-	r.cloneDir = destinationDir
+		logger.Log.Debugf("Cloning batch %d to %d: %v", i, batchEndIndex, currentPackageNames)
+		err = r.chroot.Run(func() (err error) {
+			var chrootErr error
+			// Consider the toolchain RPMs first, then built RPMs, then the already cached, and finally all remote packages.
+			preBuilt, chrootErr = r.clonePackage(finalArgs)
```

Comment:
We will stomp on the value of `preBuilt` here. We need to move this to a map, list maybe. We use this to report if a given package is cloned from the toolchain repo.

Reply:
That's true. It's a bug we have in the current code and I didn't focus on fixing this particular issue here. Having said that, I'll take a look, if I can resolve with a minor adjustemnt. Otherwise, I'll leave it as is.

Reply:
This should be good for now.

Reply:
I've made the slight change of saying that `preBuilt` means **ALL** packages were pre-built. This is consistent with the current behaviour. I wasn't sure which way we may want to go for the actual fix and it seemed too risky and robust regardless of the direction to add it to this PR.